File: java/com/tetris/Board.java
package com.tetris;

import java.util.Arrays;


public class Board {
    private Grid grid;
    private int width;
    private int height;
    private int[] widths;
    private int[] heights;
    private Grid xGrid;
    private int[] xWidths;
    private int[] xHeights;
    private boolean isCommitted = true;
    private int largestRow = 0;
    private int xLargestRow = 0;

    public Board(int width, int height) {
        this.width = width;
        this.height = height;
        grid = new Grid(width, height);
        widths = new int[height];
        heights = new int[width];
        xGrid = new Grid(width, height);
        xWidths = new int[height];
        xHeights = new int[width];
    }

    private void backup() {
        xGrid = grid.clone();
        xWidths = Arrays.copyOf(widths, widths.length);
        xHeights = Arrays.copyOf(heights, heights.length);
        xLargestRow = largestRow;
    }

    public String toString() {
        return grid.toString();
    }

    public int getWidth() {
        return width;
    }

    public int getHeight() {
        return height;
    }

    public int[] getHeights() {
        return heights;
    }

    public int[] getWidths() {
        return widths;
    }

    public void commit() {
        this.isCommitted = true;
    }

    public void commit(boolean isCommitted) {
        this.isCommitted = isCommitted;
    }

    public void undo() {
        grid = xGrid;
        widths = xWidths;
        heights = xHeights;
        largestRow = xLargestRow;
        commit();
    }

    public int dropHeight(Piece piece, int x) {
        int maxY = 0;
        int[] skirt = piece.getSkirt();
        int width = piece.getWidth();

        for (int i = 0; i < width; i++) {
            int lowestY = skirt[i];
            int comparableY = heights[x + i] - lowestY;
            maxY = Math.max(maxY, comparableY);
        }

        return maxY;
    }

    public Message insert(Piece piece, int x, int y) {
        if (!isCommitted) {
            throw new RuntimeException("Previous round has not been committed");
        }

        commit(false);
        backup();
        String color = piece.getColor().toUpperCase();
        Point[] body = piece.getBody();

        for (int i = 0; i < body.length; i++) {
            int pX = x + body[i].getX();
            int pY = y + body[i].getY();

            if (0 > pX || pX >= width || 0 > pY || pY >= height) {
                return Message.PLACE_OUT_BOUNDS;
            }

            if (grid.getCell(pX, pY) != grid.emptyCell) {
                return Message.PLACE_BAD;
            }

            widths[pY] += 1;
            heights[pX] = Math.max(heights[pX], pY + 1);
            largestRow = Math.max(largestRow, pY + 1);

            grid.set(pX, pY, color);
        }

        return Message.PLACE_OK;
    }

    private void shiftRow(int row) {
        for (int y = row + 1; y < height; y++) {
            for (int x = 0; x < width; x++) {
                String cell = grid.getCell(x, y);
                grid.setCell(x, y - 1, cell);
            }

            widths[y - 1] = widths[y];
        }
    }

    private void clearRow(int row) {
        for (int x = 0; x < width; x++) {
            heights[x] -= 1;
            grid.setCell(x, row, grid.emptyCell);
        }
    }

    public int clearRows() {
        int lineCount = 0;
        int y = 0;

        while (y < largestRow) {
            if (widths[y] != width) {
                y += 1;
                continue;
            }

            clearRow(y);
            shiftRow(y);

            lineCount += 1;
            largestRow -= 1;
        }

        return lineCount;
    }
}

File: python/lib/board.py
from math import floor, ceil
from copy import deepcopy
from utils import matrix
from .shape import Shape
from .point import Point
from .grid import Grid

class Board():
    """Representation of a Tetris board.

    Attributes:
        grid: type:Grid
        width: Board width - type:int
        height: Board height - type:int
        widths: List representing the largest width for each row - type:list
        heights: List representing the largest height for each column - type:list
        largest_row: Index of the highest row with a filled block - type:int
        is_committed: Current state of the board - type:bool
        grid_backup: Grid backup - type:Grid
        widths_backup: Widths backup - type:list
        heights_backup: Heights backup - type:list
    """

    PLACE_OK = "PLACE_OK"
    PLACE_OUT_BOUNDS = "PLACE_OUT_BOUNDS"
    PLACE_BAD = "PLACE_BAD"
    PLACE_ROW_FILLED = "PLACE_ROW_FILLED"

    @staticmethod
    def dispatch(message, *payload):
        return (message, payload)

    def __init__(self, width, height):
        self.grid = Grid(width, height, ".")
        self.width = width
        self.height = height
        self.widths = [0] * height
        self.heights = [0] * width
        self.largest_row = 0
        self.is_committed = True
        self.grid_backup = None
        self.widths_backup = None
        self.heights_backup = None

    def __str__(self):
        return str(self.grid)

    def replace(self, matrix):
        """Replaces the current board with a new board."""
        self.grid.replace(matrix)
        self.widths = [0] * self.height
        self.heights = [0] * self.width
        self.count_blocks()

    def insert(self, shape, x, y, color=None):
        """Attempts to place a shape at the specified coordinates."""
        color = color or shape.color.upper()

        if not self.is_committed:
            raise Exception("Previous round has not been committed")

        self.is_committed = False
        self.backup()

        message = self.PLACE_OK
        heights = [0] * self.width

        for coords in shape.body:
            px, py = coords + (x, y)

            if not 0 <= px < self.width or not 0 <= py < self.height:
                message = self.PLACE_OUT_BOUNDS
                break

            if self.grid.get(px, py) != ".":
                message = self.PLACE_BAD
                break

            self.grid.set(px, py, color)
            self.widths[py] += 1

            adjusted_y = 0 if coords.y == 1 else 1
            if heights[px] < adjusted_y + 1:
                heights[px] = adjusted_y + 1

            if self.widths[py] == self.width:
                message = self.PLACE_ROW_FILLED

        for i in range(len(heights)):
            self.heights[i] += heights[i]

        return self.dispatch(message, shape, x, y)

    def count_blocks(self):
        """Sets new values for largest_row, heights, and widths."""
        self.largest_row = 0
        for i in range(self.height):
            row = self.grid.get_row(i)
            should_increment_row = True

            for j in range(self.width):
                if row[j] != ".":
                    self.heights[j] += 1
                else:
                    should_increment_row = False

            if should_increment_row:
                self.widths[i] += 1
                if self.largest_row < i:
                    self.largest_row = i

    def drop_height(self, shape, x):
        """Calculates the y value where the origin of a piece will come to rest if dropped."""
        max_y = 0

        for i in range(len(shape.skirt)):
            offset_x = shape.body[i].x + x
            y = self.heights[offset_x] - shape.height - shape.skirt[i]
            if y > max_y:
                max_y = y

        return self.height - max_y - shape.height

    def clear_rows(self):
        """Clears filled rows and shifts things down."""
        line_count = 0
        for i in (range(self.largest_row + 1)):
            if any([b == "." for b in self.grid.get_row(i)]):
                continue

            for c in self.heights:
                c -= 1

            self.widths[i] = 0
            self.grid.set_row(i, ["."] * self.width)
            line_count += 1

        return line_count

    def backup(self):
        """Creates a backup of the state of the current board."""
        self.grid_backup = self.grid.clone()
        self.widths_backup = deepcopy(self.widths)
        self.heights_backup = deepcopy(self.heights)

    def commit(self):
        """Commits the current state of the board."""
        self.is_committed = True

    # def place_shape(self):
    #     self.grid.save_tmp()

    def undo(self):
        """Replaces the current state of the board with the previous state."""
        self.grid = self.grid_backup
        self.widths = self.widths_backup
        self.heights = self.heights_backup
        self.commit()

    def clear(self):
        """Resets the board."""
        self.grid.clear()
